streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -174,12 +181,11 @@ public void processInOrder(final K key, final V value, final long timestamp) {
                     }
                 }
             }
-
             //create right window for previous record
             if (latestLeftTypeWindow != null) {
-                final long leftWindowEnd = latestLeftTypeWindow.key.window().end();
-                final long rightWinStart = leftWindowEnd == windows.timeDifferenceMs() ? latestLeftTypeWindow.value.timestamp() + 1 : leftWindowEnd + 1;
-                if (!windowStartTimes.contains(rightWinStart)) {
+                final long previousRecord = latestLeftTypeWindow.key.window().end();
+                final long rightWinStart = previousRecord == windows.timeDifferenceMs() ? latestLeftTypeWindow.value.timestamp() + 1 : previousRecord + 1;
```

Comment:
If previous record is at `timeDifferenceMs` then the "left type window" we found was the window from [0,timeDifference]. Since we know it's the left window from `windowEnd == timeDifference`, we know that either we'll create a right window for the max record in the combined window, or we'll create a right window for the record at the end point of the window (which would still be the max record). 

I see that `previousRecord` is misleading, I renamed it to `previousWindowEnd`. `previousRecord` here was really just the `windowEnd`, since there might not be a _record_ at the end of the combined window, which is the case this is covering. This case is specifically for records that come outside of the combined window, but still need to create a right window for something _in_ the combined window, in which case the right window won't already exist.

Reply:
Ok before I continue to try and wrap my head around this, is this particular line going to be moot if we start tracking `previousRecordTimestamp` instead of `latestLeftTypeWindow`? The point of it was just to distinguish that special case, but now we can just say `rightWindowStart = previousRecordTimestamp + 1` -- does that sound right?